File: src/components/ai/ai-api.ts
import { getCsrfToken } from "../../lib/spira/auth";

const AI_BASE = "/api/ai";

/**
 * Headers for a mutating request (POST/PATCH/DELETE). The backend enforces CSRF
 * via the double-submit cookie pattern: echo the readable XSRF-TOKEN cookie in
 * the X-XSRF-TOKEN header, or Spring Security rejects the request with 403.
 */
function mutationHeaders(
  extra?: Record<string, string>,
): Record<string, string> {
  return { "X-XSRF-TOKEN": getCsrfToken(), ...extra };
}

/**
 * Turns a failed `Response` into a short, human-readable message — never the raw
 * body. The backend returns Spring **ProblemDetail** JSON (and Bean-Validation
 * errors) on failure; dumping that JSON into a toast is meaningless to a user, so
 * we pull out a readable field when there is one and otherwise map by status.
 *
 * - Validation errors (`400`/`422`) surface their `detail`/field message (e.g.
 *   "apiKey must be between 8 and 512 characters") — these are already written
 *   for humans.
 * - `401`/`403` → a sign-in / permission hint.
 * - `5xx` → a generic "try again" (the ProblemDetail `detail` for a 500 is just
 *   "Something went wrong… Reference: <uuid>", which we deliberately hide).
 * - A plain-text (non-JSON) body is safe to show as-is.
 */
async function friendlyError(res: Response, fallback: string): Promise<string> {
  let body = "";
  try {
    body = await res.text();
  } catch {
    /* body already consumed / unavailable */
  }

  const trimmed = body.trim();
  if (trimmed) {
    if (trimmed.startsWith("{") || trimmed.startsWith("[")) {
      try {
        const j = JSON.parse(trimmed) as {
          detail?: string;
          message?: string;
          errors?: Array<{ defaultMessage?: string }>;
          status?: number;
        };
        const fieldMsg = Array.isArray(j.errors) && j.errors[0]?.defaultMessage;
        // Only trust `detail`/`message` for non-5xx — a 500's detail carries an
        // internal reference id, not something the user can act on.
        const problemMsg = res.status < 500 ? j.detail || j.message : undefined;
        const msg = fieldMsg || problemMsg;
        if (msg) return msg;
      } catch {
        /* not valid JSON — fall through to status mapping */
      }
    } else {
      // Plain-text error body — safe to show directly.
      return trimmed;
    }
  }

  if (res.status === 401 || res.status === 403)
    return "Your session expired. Please sign in again.";
  if (res.status === 400 || res.status === 422) return fallback;
  if (res.status >= 500)
    return "Something went wrong on the server. Please try again.";
  return fallback;
}

export type HistoryEntry = { role: "user" | "assistant"; content: string };

/**
 * A file attached directly to a chat message (BUG-017): an image, PDF, or DOCX.
 * `dataUrl` is a `data:<mime>;base64,…` URL. Ephemeral — sent with this message
 * only, never saved as a resource.
 */
export type ChatAttachment = { name: string; mime: string; dataUrl: string };

export type StreamChatParams = {
  goalId?: string;
  message: string;
  history: HistoryEntry[];
  provider?: string;
  sessionType?: "chat" | "grow";
  /** Files attached directly to this message (images/PDF/DOCX). */
  attachments?: ChatAttachment[];
  /** GROW only: session length the user picked, in minutes. */
  sessionTotalMinutes?: number;
  /** GROW only: seconds left on the timer; <= 0 tells the coach to close now. */
  sessionRemainingSeconds?: number;
  onToken: (token: string) => void;
  onProposal?: (argsJson: string) => void;
  /** Progress lines (e.g. one-time coaching-library indexing in GROW).
   *  Ephemeral UI only — never part of the transcript or history. */
  onStatus?: (msg: string) => void;
  onDone: () => void;
  onError: (msg: string) => void;
};

export async function streamChat(params: StreamChatParams): Promise<void> {
  const {
    goalId,
    message,
    history,
    provider = "ANTHROPIC",
    sessionType = "chat",
    attachments,
    sessionTotalMinutes,
    sessionRemainingSeconds,
    onToken,
    onProposal,
    onStatus,
    onDone,
    onError,
  } = params;

  let response: Response;
  try {
    response = await fetch(`${AI_BASE}/chat`, {
      method: "POST",
      credentials: "include",
      headers: mutationHeaders({ "Content-Type": "application/json" }),
      body: JSON.stringify({
        goalId: goalId ? parseInt(goalId, 10) : null,
        message,
        history,
        provider,
        sessionType,
        attachments: attachments?.length ? attachments : null,
        sessionTotalMinutes: sessionTotalMinutes ?? null,
        sessionRemainingSeconds: sessionRemainingSeconds ?? null,
      }),
    });
  } catch {
    onError("NETWORK");
    return;
  }

  if (!response.ok) {
    if (response.status === 422) {
      onError("NO_KEY");
    } else {
      onError(`Server error: ${response.status}`);
    }
    return;
  }

  const reader = response.body!.getReader();
  const decoder = new TextDecoder();
  let buffer = "";

  // SSE event being assembled. Per the spec, an event may span several
  // `data:` lines (joined with "\n") and is dispatched on a blank line.
  let eventName = "";
  let dataLines: string[] = [];
  let finished = false;

  // Returns true if the stream should stop (done/error dispatched).
  const dispatch = (): boolean => {
    if (!eventName && dataLines.length === 0) return false;
    const data = dataLines.join("\n");
    dataLines = [];
    const name = eventName;
    eventName = "";

    switch (name) {
      case "token": {
        // Tokens are JSON-encoded by the backend so they survive newlines.
        let text = data;
        try {
          text = JSON.parse(data);
        } catch {
          /* fall back to raw */
        }
        onToken(text);
        return false;
      }
      case "proposal":
        onProposal?.(data.trim());
        return false;
      case "status":
        onStatus?.(data.trim());
        return false;
      case "done":
        onDone();
        return true;
      case "error":
        onError(data.trim() || "AI service error");
        return true;
      default:
        return false;
    }
  };

  try {
    while (!finished) {
      const { done, value } = await reader.read();
      if (done) break;

      buffer += decoder.decode(value, { stream: true });
      let nl: number;
      while ((nl = buffer.indexOf("\n")) >= 0) {
        let line = buffer.slice(0, nl);
        buffer = buffer.slice(nl + 1);
        if (line.endsWith("\r")) line = line.slice(0, -1);

        if (line === "") {
          if (dispatch()) {
            finished = true;
            break;
          }
        } else if (line.startsWith("event:")) {
          eventName = line.slice(6).trim();
        } else if (line.startsWith("data:")) {
          dataLines.push(line.slice(5)); // keep value verbatim; JSON handles spaces
        }
        // lines starting with ":" (comments) or unknown fields are ignored
      }
    }
  } finally {
    // cancel() aborts the underlying connection and releases the lock.
    reader.cancel().catch(() => {});
  }

  // Flush a final event that arrived without a trailing blank line (the stream
  // can close right after an `error`/`done` event) — otherwise the error would
  // be swallowed and reported as a normal completion.
  if (!finished) {
    const dispatched = eventName || dataLines.length > 0 ? dispatch() : false;
    if (!dispatched) onDone();
  }
}

export async function saveApiKey(
  provider: string,
  apiKey: string,
  model?: string,
) {
  const res = await fetch(`${AI_BASE}/keys`, {
    method: "POST",
    credentials: "include",
    headers: mutationHeaders({ "Content-Type": "application/json" }),
    body: JSON.stringify({ provider, apiKey, model: model ?? null }),
  });
  if (!res.ok) {
    throw new Error(
      await friendlyError(
        res,
        `Couldn't save the ${provider} key. Please try again.`,
      ),
    );
  }
  return res.json();
}

export async function listApiKeys() {
  const res = await fetch(`${AI_BASE}/keys`, { credentials: "include" });
  if (!res.ok) throw new Error(`Failed to list keys: ${res.status}`);
  return res.json();
}

export async function fetchProviderModels(provider: string): Promise<string[]> {
  const res = await fetch(`${AI_BASE}/keys/${provider}/models`, {
    credentials: "include",
  });
  if (!res.ok) {
    throw new Error(
      await friendlyError(
        res,
        `Couldn't load ${provider} models — check the key is valid.`,
      ),
    );
  }
  return res.json();
}

export async function updateKeyModel(provider: string, model: string) {
  const res = await fetch(`${AI_BASE}/keys/${provider}`, {
    method: "PATCH",
    credentials: "include",
    headers: mutationHeaders({ "Content-Type": "application/json" }),
    body: JSON.stringify({ model }),
  });
  if (!res.ok) {
    throw new Error(
      await friendlyError(res, "Couldn't change the model. Please try again."),
    );
  }
  return res.json();
}

// ── Provider preference sync (cross-device) ─────────────────────────────────

/** The current user's saved chat provider (null if none chosen). Best-effort. */
export async function getAiProvider(): Promise<string | null> {
  try {
    const res = await fetch(`${AI_BASE}/preferences`, {
      credentials: "include",
    });
    if (!res.ok) return null;
    const body = (await res.json()) as { provider: string | null };
    return body.provider ?? null;
  } catch {
    return null;
  }
}

/** Persist the selected chat provider so it follows the user across devices. */
export async function saveAiProvider(provider: string): Promise<void> {
  try {
    await fetch(`${AI_BASE}/preferences`, {
      method: "PUT",
      credentials: "include",
      headers: mutationHeaders({ "Content-Type": "application/json" }),
      body: JSON.stringify({ provider }),
    });
  } catch {
    /* best-effort — local choice still applies on this device */
  }
}

// ── Chat transcript sync (cross-device) ─────────────────────────────────────

/** A stored transcript for one scope. `content` is the JSON array of messages. */
export type StoredTranscript = {
  goalId: number | null;
  content: string;
  updatedAt: string | null;
};

const transcriptUrl = (goalId?: string) =>
  `${AI_BASE}/chat/transcript${goalId ? `?goalId=${parseInt(goalId, 10)}` : ""}`;

/**
 * Fetch the current user's stored transcript for a scope (a goal, or the global
 * chat when `goalId` is omitted). Best-effort: returns null on any failure so the
 * chat still works offline from its local cache.
 */
export async function getTranscript(
  goalId?: string,
): Promise<StoredTranscript | null> {
  try {
    const res = await fetch(transcriptUrl(goalId), { credentials: "include" });
    if (!res.ok) return null;
    return (await res.json()) as StoredTranscript;
  } catch {
    return null;
  }
}

/**
 * Persist the current user's transcript for a scope (last write wins). `content`
 * is the JSON array of messages (attachment file bytes already stripped).
 * Best-effort: a failure never blocks the chat.
 */
export async function putTranscript(
  goalId: string | undefined,
  content: string,
): Promise<string | null> {
  try {
    const res = await fetch(`${AI_BASE}/chat/transcript`, {
      method: "PUT",
      credentials: "include",
      headers: mutationHeaders({ "Content-Type": "application/json" }),
      body: JSON.stringify({
        goalId: goalId ? parseInt(goalId, 10) : null,
        content,
      }),
    });
    if (!res.ok) return null;
    // Return the server's new `updatedAt` so the caller can record it and avoid
    // re-adopting its own write during polling.
    const body = (await res.json()) as StoredTranscript;
    return body.updatedAt ?? null;
  } catch {
    /* best-effort — local cache still holds the transcript */
    return null;
  }
}

/** Clear the current user's stored transcript for a scope ("New chat"). */
export async function deleteTranscript(goalId?: string): Promise<void> {
  try {
    await fetch(transcriptUrl(goalId), {
      method: "DELETE",
      credentials: "include",
      headers: mutationHeaders(),
    });
  } catch {
    /* best-effort */
  }
}

// ── GROW session memory ─────────────────────────────────────────────────────

/**
 * Persist a GROW session summary on the goal ("Save memory" on the end card).
 * Later GROW sessions read it back so the coach continues the thread.
 */
export async function saveSessionMemory(
  goalId: string,
  summary: string,
): Promise<void> {
  const res = await fetch(`${AI_BASE}/goals/${parseInt(goalId, 10)}/memory`, {
    method: "POST",
    credentials: "include",
    headers: mutationHeaders({ "Content-Type": "application/json" }),
    body: JSON.stringify({ summary }),
  });
  if (!res.ok) {
    throw new Error(
      await friendlyError(
        res,
        "Couldn't save the session memory. Please try again.",
      ),
    );
  }
}

// ── Proposals ───────────────────────────────────────────────────────────────

/** A proposal as persisted server-side (status is the source of truth). */
export type ServerProposal = {
  id: number;
  goalId: number | null;
  type: string;
  payload: string; // the propose_goal_change arguments JSON (string)
  status: "PENDING" | "APPROVED" | "REJECTED";
  createdAt: string;
};

/** Fetch pending proposals for a goal (used to restore cards after a reload). */
export async function listGoalProposals(
  goalId: string,
): Promise<ServerProposal[]> {
  const res = await fetch(`${AI_BASE}/proposals/goal/${goalId}`, {
    credentials: "include",
  });
  if (!res.ok) throw new Error(`Failed to load proposals: ${res.status}`);
  return res.json();
}

/** Mark a proposal approved on the server. Best-effort: caller ignores failures. */
export async function approveProposal(id: number): Promise<void> {
  await fetch(`${AI_BASE}/proposals/${id}/approve`, {
    method: "POST",
    credentials: "include",
    headers: mutationHeaders(),
  });
}

/** Mark a proposal rejected on the server. Best-effort: caller ignores failures. */
export async function rejectProposal(id: number): Promise<void> {
  await fetch(`${AI_BASE}/proposals/${id}/reject`, {
    method: "POST",
    credentials: "include",
    headers: mutationHeaders(),
  });
}

File: android/app/src/main/java/com/spiramindscape/android/data/ai/AiApi.kt
package com.spiramindscape.android.data.ai

import com.spiramindscape.android.data.net.Network
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.channels.awaitClose
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.callbackFlow
import kotlinx.coroutines.flow.flowOn
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import org.json.JSONArray
import org.json.JSONObject
import java.util.concurrent.TimeUnit

/**
 * The AI assistant's HTTP API — the Android twin of the web `src/components/ai/ai-api.ts`,
 * talking to the same `/api/ai` endpoints on the same session cookie.
 *
 * Streaming uses Server-Sent Events exactly as the web does: `event: token|proposal|status|
 * done|error` with a JSON-encoded `data:` payload, dispatched on a blank line.
 */
object AiApi {

    private val JSON = "application/json; charset=utf-8".toMediaType()

    private const val BASE = "/api/ai"

    /**
     * A separate client for the chat stream: OkHttp's default 10s read timeout would cut a long
     * answer mid-sentence, and a streamed response has no natural "response received" moment.
     * Everything else (the cookie jar, the CSRF header) is inherited from the shared client.
     */
    private val streamClient: OkHttpClient by lazy {
        Network.okHttp.newBuilder()
            .readTimeout(0, TimeUnit.MILLISECONDS)
            .callTimeout(0, TimeUnit.MILLISECONDS)
            .build()
    }

    private fun url(path: String) = "${Network.baseUrl}$BASE$path"

    // ── Chat streaming ──────────────────────────────────────────────────────

    /** One turn of the conversation as the model should see it. */
    data class HistoryEntry(val role: String, val content: String)

    /**
     * A file attached directly to a chat message: an image, PDF or DOCX. [dataUrl] is a
     * `data:<mime>;base64,…` URL. Ephemeral — sent with this message only, never saved as a
     * resource.
     */
    data class ChatAttachment(val name: String, val mime: String, val dataUrl: String)

    /** What the stream emits. The flow completes after [Done] or [Error]. */
    sealed interface ChatEvent {
        /** A chunk of the answer. Tokens arrive in order and are appended verbatim. */
        data class Token(val text: String) : ChatEvent

        /** A `propose_goal_change` tool call, as its raw arguments JSON. */
        data class Proposal(val argsJson: String) : ChatEvent

        /** Progress the user should see but which never joins the transcript. */
        data class Status(val message: String) : ChatEvent

        data object Done : ChatEvent

        data class Error(val message: String) : ChatEvent
    }

    /** `NO_KEY` is reported as its own error so the UI can offer to add one. */
    const val ERROR_NO_KEY = "NO_KEY"
    const val ERROR_NETWORK = "NETWORK"

    /**
     * Stream one reply. Cancelling the collecting coroutine cancels the HTTP call, which is how
     * the composer's Stop button works.
     */
    fun streamChat(
        goalId: String? = null,
        message: String,
        history: List<HistoryEntry>,
        provider: String = "ANTHROPIC",
        sessionType: String = "chat",
        attachments: List<ChatAttachment> = emptyList(),
        sessionTotalMinutes: Int? = null,
        sessionRemainingSeconds: Int? = null,
    ): Flow<ChatEvent> = callbackFlow {
        val body = JSONObject().apply {
            put("goalId", goalId?.toLongOrNull() ?: JSONObject.NULL)
            put("message", message)
            put(
                "history",
                JSONArray().apply {
                    history.forEach {
                        put(JSONObject().put("role", it.role).put("content", it.content))
                    }
                },
            )
            put("provider", provider)
            put("sessionType", sessionType)
            put(
                "attachments",
                if (attachments.isEmpty()) {
                    JSONObject.NULL
                } else {
                    JSONArray().apply {
                        attachments.forEach {
                            put(
                                JSONObject()
                                    .put("name", it.name)
                                    .put("mime", it.mime)
                                    .put("dataUrl", it.dataUrl),
                            )
                        }
                    }
                },
            )
            put("sessionTotalMinutes", sessionTotalMinutes ?: JSONObject.NULL)
            put("sessionRemainingSeconds", sessionRemainingSeconds ?: JSONObject.NULL)
        }

        val call = streamClient.newCall(
            Request.Builder()
                .url(url("/chat"))
                .header("Accept", "text/event-stream")
                .post(body.toString().toRequestBody(JSON))
                .build(),
        )

        try {
            val response = try {
                call.execute()
            } catch (e: Exception) {
                trySend(ChatEvent.Error(ERROR_NETWORK))
                close()
                return@callbackFlow
            }

            response.use { res ->
                if (!res.isSuccessful) {
                    // 422 is the backend's "no API key saved for this provider".
                    trySend(
                        ChatEvent.Error(
                            if (res.code == 422) ERROR_NO_KEY else "Server error: ${res.code}",
                        ),
                    )
                    close()
                    return@callbackFlow
                }

                val source = res.body?.source()
                if (source == null) {
                    trySend(ChatEvent.Error(ERROR_NETWORK))
                    close()
                    return@callbackFlow
                }

                var eventName = ""
                val dataLines = mutableListOf<String>()

                /** Returns true once a terminal event (done/error) has been dispatched. */
                fun dispatch(): Boolean {
                    if (eventName.isEmpty() && dataLines.isEmpty()) return false
                    val data = dataLines.joinToString("\n")
                    dataLines.clear()
                    val name = eventName
                    eventName = ""
                    return when (name) {
                        // Tokens are JSON-encoded by the backend so they survive newlines.
                        "token" -> { trySend(ChatEvent.Token(decodeToken(data))); false }
                        "proposal" -> { trySend(ChatEvent.Proposal(data.trim())); false }
                        "status" -> { trySend(ChatEvent.Status(data.trim())); false }
                        "done" -> { trySend(ChatEvent.Done); true }
                        "error" -> {
                            trySend(ChatEvent.Error(data.trim().ifEmpty { "AI service error" }))
                            true
                        }
                        else -> false
                    }
                }

                var finished = false
                while (!finished) {
                    val line = source.readUtf8Line() ?: break
                    when {
                        line.isEmpty() -> if (dispatch()) finished = true
                        line.startsWith("event:") -> eventName = line.substring(6).trim()
                        // Keep the value verbatim — JSON decoding handles leading spaces.
                        line.startsWith("data:") -> dataLines += line.substring(5)
                        // ":" comments and unknown fields are ignored.
                    }
                }

                // The stream can close right after an event without a trailing blank line;
                // flush it, or an error would be swallowed and read as a normal completion.
                if (!finished) {
                    val dispatched = if (eventName.isNotEmpty() || dataLines.isNotEmpty()) dispatch() else false
                    if (!dispatched) trySend(ChatEvent.Done)
                }
            }
        } catch (e: Exception) {
            trySend(ChatEvent.Error(ERROR_NETWORK))
        }
        close()

        awaitClose { call.cancel() }
    }.flowOn(Dispatchers.IO)

    /** A token arrives JSON-encoded; fall back to the raw text if it isn't valid JSON. */
    private fun decodeToken(data: String): String =
        runCatching { JSONArray("[$data]").getString(0) }.getOrDefault(data)

    // ── Keys and models ─────────────────────────────────────────────────────

    /** A saved key as the server describes it — never the key itself. */
    data class KeyInfo(val provider: String, val hint: String?, val model: String?)

    suspend fun listKeys(): List<KeyInfo> = io {
        val text = getText(url("/keys")) ?: return@io emptyList()
        val arr = JSONArray(text)
        (0 until arr.length()).map { i ->
            val o = arr.getJSONObject(i)
            KeyInfo(
                provider = o.optString("provider"),
                hint = o.optStringOrNull("hint"),
                model = o.optStringOrNull("model"),
            )
        }
    }

    suspend fun saveKey(provider: String, apiKey: String, model: String? = null): KeyInfo = io {
        val body = JSONObject()
            .put("provider", provider)
            .put("apiKey", apiKey)
            .put("model", model ?: JSONObject.NULL)
        val text = postText(url("/keys"), body)
            ?: throw AiException("Couldn't save the $provider key. Please try again.")
        val o = JSONObject(text)
        KeyInfo(o.optString("provider"), o.optStringOrNull("hint"), o.optStringOrNull("model"))
    }

    suspend fun listProviderModels(provider: String): List<String> = io {
        val text = getText(url("/keys/$provider/models"))
            ?: throw AiException("Couldn't load $provider models — check the key is valid.")
        val arr = JSONArray(text)
        (0 until arr.length()).map { arr.getString(it) }
    }

    suspend fun updateKeyModel(provider: String, model: String): Unit = io {
        val request = Request.Builder()
            .url(url("/keys/$provider"))
            .patch(JSONObject().put("model", model).toString().toRequestBody(JSON))
            .build()
        Network.okHttp.newCall(request).execute().use { res ->
            if (!res.isSuccessful) throw AiException(friendlyError(res.code, res.body?.string()))
        }
    }

    suspend fun deleteKey(provider: String): Unit = io {
        val request = Request.Builder().url(url("/keys/$provider")).delete().build()
        Network.okHttp.newCall(request).execute().use { }
    }

    // ── Provider preference (follows the user across devices) ───────────────

    suspend fun getProvider(): String? = io {
        val text = getText(url("/preferences")) ?: return@io null
        JSONObject(text).optStringOrNull("provider")
    }

    suspend fun saveProvider(provider: String): Unit = io {
        runCatching { putText(url("/preferences"), JSONObject().put("provider", provider)) }
        Unit
    }

    // ── Transcript sync ─────────────────────────────────────────────────────

    /** A stored transcript for one scope; [content] is the JSON array of messages. */
    data class StoredTranscript(val content: String, val updatedAt: String?)

    private fun transcriptUrl(goalId: String?) =
        url("/chat/transcript") + if (goalId != null) "?goalId=${goalId.toLongOrNull() ?: 0}" else ""

    /** Best-effort: null on any failure, so the chat still works from its local cache. */
    suspend fun getTranscript(goalId: String?): StoredTranscript? = io {
        val text = getText(transcriptUrl(goalId)) ?: return@io null
        val o = JSONObject(text)
        StoredTranscript(o.optString("content"), o.optStringOrNull("updatedAt"))
    }

    /** Last write wins. Returns the server's new `updatedAt`, or null if the write failed. */
    suspend fun putTranscript(goalId: String?, content: String): String? = io {
        val body = JSONObject()
            .put("goalId", goalId?.toLongOrNull() ?: JSONObject.NULL)
            .put("content", content)
        val text = runCatching { putText(url("/chat/transcript"), body) }.getOrNull() ?: return@io null
        JSONObject(text).optStringOrNull("updatedAt")
    }

    /** Clear the stored transcript for a scope ("New chat"). */
    suspend fun deleteTranscript(goalId: String?): Unit = io {
        val request = Request.Builder().url(transcriptUrl(goalId)).delete().build()
        runCatching { Network.okHttp.newCall(request).execute().use { } }
        Unit
    }

    // ── GROW session memory ─────────────────────────────────────────────────

    /** Persist a GROW session summary on the goal, so later sessions continue the thread. */
    suspend fun saveGoalMemory(goalId: String, summary: String): Unit = io {
        val body = JSONObject().put("summary", summary)
        postText(url("/goals/${goalId.toLongOrNull() ?: 0}/memory"), body)
            ?: throw AiException("Couldn't save the session memory. Please try again.")
        Unit
    }

    // ── Proposals ───────────────────────────────────────────────────────────

    /** A proposal as persisted server-side; its status is the source of truth. */
    data class ServerProposal(
        val id: Long,
        val goalId: Long?,
        val type: String,
        val payload: String,
        val status: String,
    )

    suspend fun listGoalProposals(goalId: String): List<ServerProposal> = io {
        val text = getText(url("/proposals/goal/$goalId")) ?: return@io emptyList()
        val arr = JSONArray(text)
        (0 until arr.length()).map { i ->
            val o = arr.getJSONObject(i)
            ServerProposal(
                id = o.optLong("id"),
                goalId = if (o.isNull("goalId")) null else o.optLong("goalId"),
                type = o.optString("type"),
                payload = o.optString("payload"),
                status = o.optString("status"),
            )
        }
    }

    suspend fun approveProposal(id: Long): Unit = io { postText(url("/proposals/$id/approve"), null); Unit }

    suspend fun rejectProposal(id: Long): Unit = io { postText(url("/proposals/$id/reject"), null); Unit }

    // ── plumbing ────────────────────────────────────────────────────────────

    /** Raised when a call fails with a message worth showing the user. */
    class AiException(message: String) : Exception(message)

    private suspend fun <T> io(block: suspend () -> T): T =
        kotlinx.coroutines.withContext(Dispatchers.IO) { block() }

    private fun getText(fullUrl: String): String? =
        Network.okHttp.newCall(Request.Builder().url(fullUrl).build()).execute().use { res ->
            if (res.isSuccessful) res.body?.string() else null
        }

    private fun postText(fullUrl: String, body: JSONObject?): String? {
        val request = Request.Builder()
            .url(fullUrl)
            .post((body?.toString() ?: "{}").toRequestBody(JSON))
            .build()
        return Network.okHttp.newCall(request).execute().use { res ->
            if (res.isSuccessful) res.body?.string() else null
        }
    }

    private fun putText(fullUrl: String, body: JSONObject): String? {
        val request = Request.Builder()
            .url(fullUrl)
            .put(body.toString().toRequestBody(JSON))
            .build()
        return Network.okHttp.newCall(request).execute().use { res ->
            if (res.isSuccessful) res.body?.string() else null
        }
    }

    /**
     * A failed response as one readable sentence — never the raw ProblemDetail JSON. Validation
     * messages are already written for humans; a 5xx `detail` is only an internal reference id,
     * so it is deliberately hidden.
     */
    internal fun friendlyError(status: Int, body: String?): String {
        val trimmed = body?.trim().orEmpty()
        if (trimmed.isNotEmpty()) {
            if (trimmed.startsWith("{") || trimmed.startsWith("[")) {
                runCatching {
                    val o = JSONObject(trimmed)
                    val fieldMsg = o.optJSONArray("errors")
                        ?.optJSONObject(0)
                        ?.optStringOrNull("defaultMessage")
                    val problemMsg = if (status < 500) {
                        o.optStringOrNull("detail") ?: o.optStringOrNull("message")
                    } else {
                        null
                    }
                    (fieldMsg ?: problemMsg)?.let { return it }
                }
            } else {
                return trimmed
            }
        }
        return when {
            status == 401 || status == 403 -> "Your session expired. Please sign in again."
            status >= 500 -> "Something went wrong on the server. Please try again."
            else -> "That didn't work. Please try again."
        }
    }
}

/** `optString` returns "" for a JSON null, which is rarely what a nullable field wants. */
internal fun JSONObject.optStringOrNull(key: String): String? =
    if (isNull(key)) null else optString(key).takeIf { it.isNotEmpty() }
